### PR TITLE
feat(composer)!: Add chain_id check on executor build

### PR DESCRIPTION
**Pull request by eoroshiba**
## Summary
Added check to ensure the executor's configured chain_id and the sequencer client's actual chain_id match.

## Background
This check ensures that the composer is communicating with the sequencer RPC on the correct network.

## Changes
Added chain_id check to the composer executor builder.

Made executor builder async to accommodate for async communication with sequencer client.

Moved nonce guard declaration from mock startup to individual tests in executor tests.

Added function to mount a status response with a given chain_id to the mock. This is so the executor builder can gather the status response without throwing an HTTP 404 error.

## Testing
Added executor test to validate the failure of an executor build with a chain_id mismatch.

## Related Issues
Part of #986

### CODE REVIEW

File: crates/astria-composer/src/executor/builder.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/builder.rs:41]: Please don't make the builder async. If one needs config-like input, then move this into the `run`/`run_until_stopped` function and run it with a retry.

File: crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs
Comment (SuperFluffy) [inline crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs:82]: Please construct this object ad-hoc instead of a fixed string.

File: crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs
Comment (SuperFluffy) [inline crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs:66]: No need to return the mock guard into the builder. There isn't actually something to assert on other that the mock guard was triggered. Just mount it non-scoped and don't return a guard.

**Review by SuperFluffy — CHANGES_REQUESTED** [review]

File: crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs
Comment (eoroshiba) [reply crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs]: Since status::Response has no default configuration, I'm not sure how to get around doing an fixed declaration of the response somewhere. As of now, I've just moved the construction of the response inside the mount function. Also worth noting that this is copied from the status response mount for the sequencer relayer, so any change made here may need to be reflected there as well. Do you have a preferred solution in mind for the fixed strings?

File: crates/astria-composer/src/composer.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/composer.rs:243]: This performs a check but there is no actual effect on the rest of the code. Effectively, this just performs a get and wraps the return value into a `Result`. You need to have some more logic (like returning early) to have an effect.

Having said that, move this into `Executor::run_until_stopped`. The composer task acts more like a collection of sub tasks, while the sub tasks should be self-contained (including verification of chain IDs).

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:428]: This future should be wrapped in a retry because composer is likely to start way earlier then sequencer and cometbft. This means that you will a bunch of crash-look-backoff errors in kubernetes.

See conductor for doing essentially the same thing: https://github.com/astriaorg/astria/blob/c31c10b6db41b19fdce8e4b172363f855d88be23/crates/astria-conductor/src/celestia/mod.rs#L645

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:431]: You can avoid the clones/allocations. The `Genesis` return value has a `chain_id` field that provides a `as_str()` method:
https://docs.rs/tendermint/0.37.0/tendermint/genesis/struct.Genesis.html
https://docs.rs/tendermint/0.37.0/tendermint/chain/id/struct.Id.html#method.as_str

The check below would then become:
```rust
ensure!(
    &self.sequencer_chain_id == genesis.chain_id.as_str(), "<text>"
);
```

File: crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs
Comment (SuperFluffy) [reply crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs]: You just have to manually fill in the blanks. :)

If you use the genesis RPC instead of status (which I think is more reasonable), then you can just copy what we have in conductor:
https://github.com/astriaorg/astria/blob/c31c10b6db41b19fdce8e4b172363f855d88be23/crates/astria-conductor/tests/blackbox/helpers/mod.rs#L287

Comment (eoroshiba) [reply crates/astria-composer/tests/blackbox/helper/mock_sequencer.rs]: Should be resolved in #[f3c96ff]. Thanks for all the feedback :)

File: crates/astria-composer/src/composer.rs
Comment (eoroshiba) [reply crates/astria-composer/src/composer.rs]: Resolved in f3c96ff

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Resolved in f3c96ff.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Done in f3c96ff.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Resolved in f3c96ff.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Done in f3c96ff.

File: crates/astria-composer/src/executor/builder.rs
Comment (joroshiba) [inline crates/astria-composer/src/executor/builder.rs:18]: unclear as to why this change is here to me

File: crates/astria-composer/src/executor/tests.rs
Comment (joroshiba) [inline crates/astria-composer/src/executor/tests.rs:482]: remove

File: crates/astria-composer/src/executor/tests.rs
Comment (joroshiba) [inline crates/astria-composer/src/executor/tests.rs:479]: nit: this is repeated a number of times perhaps a helper that does exactly this `mount_default_nonce_query_guard` function?

File: crates/astria-composer/src/executor/mod.rs
Comment (joroshiba) [inline crates/astria-composer/src/executor/mod.rs:455]: nit: this is accurate but concerned the message might not give enough context information to operator to understand how configuration is wrong.

Something like: 
`mismatch in chain id: config specifies {0}, but sequencer rpc is for {1}`.

File: crates/astria-composer/src/executor/mod.rs
Comment (joroshiba) [inline crates/astria-composer/src/executor/mod.rs:428]: perhaps simplify with `get_sequencer_chain_id` and migrate the ensure to the `run_until_stopped` the `get_sequencer_chain_id` function could also not be a method taking self but a function taking just a sequencer client (see `get_latest_nonce` as example) ?

The function title is a reason this can fail, but it can also fail for errors in grabbing genesis, feels like the function has too many responsibilities to me, don't think the ensure necessarily needs to be it's own method if its "grab the data needed then ensure it's good"

File: crates/astria-composer/src/executor/builder.rs
Comment (joroshiba) [inline crates/astria-composer/src/executor/builder.rs:51]: nit: don't think this file needs any changes?

File: crates/astria-composer/src/executor/mod.rs
Comment (joroshiba) [inline crates/astria-composer/src/executor/mod.rs:199]: I don't believe this needs to be assigned to anything as written, but bigger note down on mild update which can be made

File: crates/astria-composer/src/executor/builder.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/builder.rs:18]: Should be resolved with be387af.

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs]: Done in be387af.

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs]: Done in be387af.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Done in be387af.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Done in b8d148f.

File: crates/astria-composer/src/executor/builder.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/builder.rs]: Fixed in b8d148f.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Resolved in b8d148f.

File: crates/astria-composer/src/executor/builder.rs
Comment (SuperFluffy) [reply crates/astria-composer/src/executor/builder.rs:15]: Undo the change, use `WrapErr as _`

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:206]: IMO, factor this into a method `async fn setup` or `async fn init` or `async fn pre_run_checks` and perform the checks there.

From inside `run_until_stopped` you can then do something like below so that your task is a good citizen and doesn't just ignore the shutdown signal during setup.


```rust
select!(
    biased;
    () = self.shutdown_token.cancelled() => {
        info!("received shutdown signal while running initialization routines; exiting");
        return Ok(());
    }

    res = self.setup() => {
        if let Err(e) = res {
            bail!("runtime initialization failed");
        }
);
```

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:435]: This comment is wrong. You are not performing that check in here.

Also, if you want to add doc comments, use triple backticks: `/// <doc-string>`

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:461]: The conversion is unnecessary, just return `genesis.chain_id`

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/tests.rs:190]: For a followup: Sucks that we need this here. We should just rip out all of these wrongly placed pseudo-blackbox tests and just rely on `tests/blackbox`.

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:211]: You `remote_chain_id.as_str()` from https://docs.rs/tendermint/latest/tendermint/chain/id/struct.Id.html (important due to my suggestion to just return `tendermint::chain::Id` withot allocating).

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:212]: You don't need the numbers, and you can rely on interpolation (also I recommend backticks to make it easier for a human to parse the text). I also suggest to remove mention of "config" - it's not not wrong, but this part of the code has no concept of the actual config, so I'd leave it more generic:
```suggestion
            "mismatch in Sequencer chain IDs: expected `{}`, but Sequencer returned `{remote_chain_id}`",
            self.sequencer_chain_id,
```

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:46]:
```suggestion
        Client as _,
```

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/tests.rs:545]: So this test is dangerous because it could silently break: somebody might break the chain-ID check in a refactor, but then subsequently add code that still causes this method to exit with an error. Then, under test conditions, everything looks peachy, even though we have a problem.

IMO the most straightforward way to fix this is:

1. refactor the check into a free-standing function like so:

```rust
#[derive(Debug, thiserror::Error)]
enum EnsureChainIdError {
    // <other error variants from getting genesis, etc>
    #[error("expected chain ID `{expected}`, but received `{actual}`)]
    WrongChainId { expected: String, actual: tendermint::chain::Id, }
}

fn ensure_chain_id_is_correct(client: tendermint::HttpClient, expected_chain_id: String) -> Result<(), EnsureChainIdError> {
    // code to get the chain ID
    if remote_chain_id != expected_chain_id {
        return Err(WrongChainId { expected: expected_chain_id, actual: remote_chain_id});
    }
    Ok(())
}
```

You can then actually check the function without setting up the rest, which makes things a bit easier, i.e. you could check it like so:
```rust
match err {
    EnsureChainIdError::WrongChainId { expected, actual } => {
        assert_eq!(expected, <configured-expected>);
        assert_eq!(actual, <configured-actual>);
    }
    other => panic!("expected `EnsureChainIdError::WrongChainId`, but got `{other}`),
}
```


If you still want to test `run_until_stopped` (like above, just some more code):

```rust
// <code as it's written right now>
let err = exector.run_until_stopped().await.expect_err("should exit with an error when reading a bad chain ID, but exited with success");

for cause in err.chain() {
    if let Some(err) = cause.downcast_ref::<EnsureChainIdError>() {
        match err {
            EnsureChainIdError::WrongChainId { expected, actual } => {
                assert_eq!(expected, <configured-expected>);
                assert_eq!(actual, <configured-actual>);
            }
            other => panic!("expected `EnsureChainIdError::WrongChainId`, but got `{other}`),
     }
}
```

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/tests.rs:380]: remove?

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/tests.rs:283]: remove?

**Review by SuperFluffy — COMMENTED** [review]
Code is moving in the right direction, thank you for bearing with me and working on it. :)

I have left a bunch of comments on what what I'd like to see changed to test this.

The tests are currently not set up in the most ideal way, but short of making the exit reason publicly observable I don't see a way for you to test it other than you did.

I would only ask for you to type out the returned errors so that the tests won't stop silently breaking (and yet still passing).

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs:461]: `chain_id` is of type `tendermint::chain::Id`. I may be mistaken, but I believe we'll have to de a conversion at some point for the comparison. Would you prefer to return the raw chain ID and do the conversion in `setup()`?

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs:198]: Would this be an issue for later or would it be better to do that in this PR?

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [reply crates/astria-composer/src/executor/mod.rs:461]: You don't need to convert at all. It provides this: https://docs.rs/tendermint/latest/tendermint/chain/id/struct.Id.html#method.as_str

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [reply crates/astria-composer/src/executor/tests.rs:198]: No, that's why I prefixed it with "for a followup" :-)

File: crates/astria-composer/src/executor/builder.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/builder.rs]: Resolved in 8523e04.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Implemented in 8523e04.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Apologies for the confusion, I saw your suggestion to compare using `as_str()` below and then deleted my previous comment. Should be resolved in 8523e04 :)

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs:199]: Ah, I see. I was confused as to whether it meant followup in this PR or not. Thanks for the clarification! I also changed the comment in 8523e04 to reflect now mounting the genesis as opposed to a status response.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Fixed in 8523e04.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Resolved in 8523e04.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Resolved in 8523e04.

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs]: Resolved in 8523e04.

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs]: Removed in 8523e04.

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs]: Removed in 8523e04.

**Comment by eoroshiba** [comment]
@SuperFluffy I should be thanking you for bearing with me! Recent commits should implement all the changes you requested.

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:471]:
```suggest
    async fn get_sequencer_chain_id(&self) -> Result<tendermint::chain::Id, sequener_client::tendermint_rpc::Error> {
```

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:108]: Commented below, remove the `FailedtoGetGenesis` variant, keep the `FailedtoGetChainId`

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:104]: remove that comment :) That was just for not typing out everything in my review.

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:106]: Don't throw away the error:
```suggestion
    GetChainId(#[source] sequencer_client::tendermint_rpc::Error),
```
^ check if that's the right type. I think it is

Note on the variant: because this already is an error it's recommended to not repeat "failed" or "error" in the variants.

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:105]:
```suggestion
    #[error("failed to obtain sequencer chain ID after multiple retries")]
```

File: crates/astria-composer/src/executor/mod.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/mod.rs:227]:
```suggestion
                res.wrap_err("required pre-run checks failed)?;
```

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/tests.rs:598]: You can remove this block. You are already testing below, so this doesn't add anything of value.

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/tests.rs:122]: To get a better error diagnostic:
```suggestion
#[track_caller]
fn assert_chain_id_err(
```

**Review by SuperFluffy — APPROVED** [review]
Looks good. Thank you very much!

I have left a few comments, primarily to not just throw away the error in `get_sequencer_chain_id`. Please address these before merging.

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs:123]: Done in 33ae83d.

File: crates/astria-composer/src/executor/tests.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/tests.rs]: Done in 33ae83d.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Done in 33ae83d.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Suggestion committed in 33ae83d.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Done in 33ae83d.

File: crates/astria-composer/src/executor/mod.rs
Comment (eoroshiba) [reply crates/astria-composer/src/executor/mod.rs]: Done in 33ae83d.

File: crates/astria-composer/src/executor/tests.rs
Comment (SuperFluffy) [inline crates/astria-composer/src/executor/tests.rs:137]: Last change before merge:
```suggestion
            panic!("expected `EnsureChainIdError::WrongChainId`, but got '{other:?}'")
```
^ I'd have done it myself but don't have push rights for your fork.

**Pull request merged**
